Feature/cas

loganfreeman opened the pull request:
add support for omniauth_cas

need to add the following setting in application.yml to enable it:

```
use_cas: true
use_cas_only: false 
```

use_cas: flag to turn on and off omniauth_cas
use_cas_only: if false will use both database and omniauth_cas

wtholt commented on app/controllers/identities/omniauth_callbacks_controller.rb:
```diff
@@ -30,5 +30,19 @@ def shibboleth
+    @identity = Identity.find_for_cas_oauth(request.env["omniauth.auth"], current_identity)
+
+    if @identity.persisted?
+      sign_in_and_redirect @identity, :event => :authentication
```
We're working on better styling our Ruby code. Reference - https://github.com/bbatsov/ruby-style-guide

Here, use new Ruby syntax instead of hash rockets, e.g.

     event: :authentication

Parentheses for method arguments

     sign_in_and_redirect(@identity ...)

loganfreeman replied:
Thanks

jleonardw9 commented on config/locales/en.yml:
```diff
@@ -1124,10 +1124,10 @@ en:
   login_password: "Password:"
   remember_user: "Remember me:"
   signin:
-    button1: "MUSC Users Click Here"
```
Please remove en.yml from your pull request. We need to keep our MUSC wording in place.

jleonardw9 commented on Gemfile:
```diff
@@ -55,6 +55,8 @@ gem 'surveyor', :git => 'https://github.com/caboteria/surveyor.git', branch: 'ra
 gem 'therubyracer', '0.10.2', :platforms => :ruby
 gem 'uglifier', '>= 1.0.3'
 gem 'will_paginate'
+gem 'omniauth-cas'
```
It looks like this gem is no longer maintained, the last contribution was 2 years ago. You may need to find another solution.
<img width="1126" alt="screen shot 2016-04-15 at 3 25 14 pm" src="https://cloud.githubusercontent.com/assets/1830708/14572731/627f3710-031e-11e6-8a56-6f87b3bfb73e.png">

rkmadsen replied:
As far as I can tell omni-cas is the only existing solution at this point. Many of the libraries that are used in SPARC are many years old. If this is the metric that code can be contributed we will be stuck in adding new solutions that use existing code that support the libraries that are used, such as omni-auth. Are you suggesting that we must write a new omni-auth plugin of our own and then get then contribute that?

rkmadsen replied:
We'll can look at possibly taking over that gem or writing a new implementation but for now would it  be possible to include it as an option for CAS auth?

sparc-request replied:
Yes, let me look at the gem in the morning. It could be that it's just fine.

On Apr 18, 2016, at 6:46 PM, rkmadsen <notifications@github.com<mailto:notifications@github.com>> wrote:




In Gemfile<https://github.com/sparc-request/sparc-request/pull/349#discussion_r60147597>:

> @@ -55,6 +55,8 @@ gem 'surveyor', :git => 'https://github.com/caboteria/surveyor.git', branch: 'ra
>  gem 'therubyracer', '0.10.2', :platforms => :ruby
>  gem 'uglifier', '>= 1.0.3'
>  gem 'will_paginate'
> +gem 'omniauth-cas'


We'll can look at possibly taking over that gem or writing a new implementation but for now would it be possible to include it as an option for CAS auth?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub<https://github.com/sparc-request/sparc-request/pull/349/files/7b28d9eaf48809083d026887bf07a4fac3575360#r60147597>

jleonardw9 replied:
I looked at the gem some more last night, and I think it looks good. So I think the only issues that need to be corrected are the Ruby styling and the customization of your own en.yml file. Check out this guide we wrote up to help you with your customizations: https://github.com/sparc-request/sparc-request/blob/master/CUSTOMIZE.md

rkmadsen replied:
Great thanks for reviewing it and having patience with us! Sorry for the confusion on the pull requests, we're just getting started on sending changes your way.

loganfreeman commented:
Sorry, my fault. I will reopen it.

loganfreeman commented:
I will correct the styling and yml file and resend the request

loganfreeman commented:
I corrected the styling issue, removed the change to en.yml. 
I added a helper to deal with the issue of sign in button text using different institution name. 

Please review

loganfreeman commented:
just found there is a local_override. removed the helper, and just use key from local file.

jwiel86 commented on app/models/identity.rb:
```diff
@@ -56,25 +55,25 @@ class Identity < ActiveRecord::Base
-  has_many :received_toast_messages, :class_name => 'ToastMessage', :foreign_key => 'to', :dependent => :destroy
-  has_many :sent_toast_messages, :class_name => 'ToastMessage', :foreign_key => 'from', :dependent => :destroy
-  has_many :notes, :dependent => :destroy
+  has_many :approvals, dependent: :destroy
```
Thank you for updating this file to follow our style guide specifications!

jwiel86 commented on config/routes.rb:
```diff
@@ -19,21 +19,28 @@
 # TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 SparcRails::Application.routes.draw do
```
Thanks again for following the style guide!  We appreciate it!

amcates commented on app/controllers/identities/omniauth_callbacks_controller.rb:
```diff
@@ -20,15 +20,26 @@
+  def cas
+    @identity = Identity.find_for_cas_oauth(request.env['omniauth.auth'], current_identity)
+
+    if !@identity.nil?
```
Shouldn't this be @identity.persisted?  Just like def shibboleth on line 25.

loganfreeman replied:
```ruby
  def self.find_for_cas_oauth(auth, _signed_in_resource = nil)
    Directory.search(auth.uid).first
  end
```

Directory.search will search and then return the first found. If it is not nil, then it is always persisted. Should check nil first. once nil checking passes, no need to check persisted. 

The logic here is that if a user successfully authenticated through CAS, then always succeed. unless this is not always true.

amcates commented:
@loganfreeman I have a couple issues with this pull request.  All of which I think can be easily resolved.  I feel like the refactoring should be done separately from the actual logic added.  It makes it much harder to determine exactly what has changed.  Second, why did you extract from obis_setup?   This isn't something that was required for your pull request.  I would prefer that pull requests be concise and easily explainable.  

With my comment about 'persisted?', the system creates new local identities once authenticated.  This is both to make the system faster (returning identities verses querying external systems) and it gives us a hard record of what the identity looked like.  

I think prior to future pull requests we should probably have a 10-15 conversation.  

Thanks,

Andrew

loganfreeman commented:
@amcates  for the obis_setup, its purpose is to load some constants. But the calling of the initializers has some order, such that the obis_setup initializer is not the first one to be called and the constants are not available for those called before obis_setup.
So I move the constants loading stuff to environement.rb.

loganfreeman commented:
@amcates I can make a new pull request without the obis_setup change. the devise initializer needs to know from application.yml what autheticattion provider to use. but it is called after obis_setup. So it needs read the application.yml again.
